File: mssp/archaeology/019-dict-get-sentinel/src/DMS/report.py
"""What a person is shown.

`value` is never printed without `present`, because that pairing is the entire
subject of the entry.
"""


def reads(rows):
    lines = ["    mapping         reader                  value    present   raised"]
    for row in rows:
        present = "-" if row.get("present") is None else str(row["present"])
        lines.append(f'    {row["mapping"]:<15} {row["reader"]:<23} {repr(row["value"]):<8} '
                     f'{present:<9} {row["raised"] or "-"}')
    return "\n".join(lines)


def falsy(rows, collapsed, present):
    lines = ["    case                        not d.get('a')   'a' in d"]
    for row in rows:
        lines.append(f'    {row["case"]:<27} {str(row["falsy"]):<16} {row["has_key"]}')
    lines.append(f"    {collapsed} of {len(rows)} take the same branch, and {present} of those "
                 f"{collapsed} have the key")
    return "\n".join(lines)

File: mssp/archaeology/020-rust-must-use/src/DMS/report.mjs
// What a person is shown.
//
// `compiles` is never printed without the warning column beside it, because
// "it built" is the answer three different situations share here.
const pad = (value, width) => String(value).padEnd(width);

export function table(rows) {
  const lines = ["    route                compiles   must_use warning   error"];
  for (const row of rows) {
    lines.push(`    ${pad(row.route.NAME, 20)} ${pad(row.compiles ? "yes" : "NO", 10)} `
      + `${pad(row.mustUseWarning ? "yes" : row.denied ? "as an ERROR" : "-", 18)} `
      + `${row.errorCode ?? "-"}`);
  }
  return lines.join("\n");
}

export function boundary(discardPlain, discardStrict, assign) {
  return [
    "  the boundary, in two facts:",
    `    extraction is forced      - ${assign.route.NAME} does not compile (${assign.errorCode})`,
    `    discarding is not         - ${discardPlain.route.NAME} compiles, no warning`,
    `    and the loudest setting available does not change that: still `
      + `${discardStrict.compiles ? "compiles" : "refused"}`,
    "",
    "  So the type system forces the Err case to be NAMED wherever the value is used,",
    "  and forces nothing at all where the value is thrown away.",
  ].join("\n");
}
